agent/engine/containermetadata/metadata_manager.go

```diff
@@ -0,0 +1,121 @@
+func NewMetadataManager(client dockerDummyClient, cfg *config.Config) MetadataManager {
+	seelog.Debug("Creating metadata manager from configuration: %s", cfg)
+	return &metadataManager{
+		client: client,
+		cfg:    cfg,
+	}
+}
+
+// SetContainerInstanceArn sets the metadataManager's ContainerInstanceArn which is not available
+// at its creation as this information is not present directly at the agent's startup
+func (manager *metadataManager) SetContainerInstanceArn(containerInstanceArn string) {
+	seelog.Debugf("Setting metadata manager's ContainerInstanceArn to %s", containerInstanceArn)
+	manager.containerInstanceArn = containerInstanceArn
+}
+
+// UpdateMetadata updates the metadata file after container starts and dynamic
+// metadata is available
+func (manager *metadataManager) UpdateMetadata(dockerID string, task *api.Task, container *api.Container) error {
+	// Do not update (non-existent) metadata file for internal containers
+	if container.IsInternal {
+		return nil
+	}
+
+	// Verify metadata file exists before proceeding
+	if !mdFileExist(task, container, manager.cfg.DataDir) {
+		expectedPath, _ := getMetadataFilePath(task, container, manager.cfg.DataDir)
+		return fmt.Errorf("Failed to update metadata file: %s does not exist", expectedPath)
+	}
+
+	// Get docker container information through api call
+	dockerContainer, err := manager.client.InspectContainer(dockerID, inspectContainerTimeout)
+	if err != nil {
+		return err
+	}
+
+	// Ensure we do not update a container that is invalid or is not running
+	if dockerContainer == nil || !dockerContainer.State.Running {
+		return fmt.Errorf("Failed to update metadata file: container not running or invalid")
+	}
+
+	// Get last metadata file creation time
+	createTime, err := getMetadataFileUpdateTime(task, container, manager.cfg.DataDir)
+	createTimeFmt := ""
+	if err != nil {
+		seelog.Errorf("Metadata file create time stamp not available: %v", err)
+	} else {
+		createTimeFmt = createTime.Format(time.StampNano)
+	}
+	updateTimeFmt := ""
+
+	// Acquire the metadata then write it in JSON format to the file
+	metadata := manager.acquireMetadata(createTimeFmt, updateTimeFmt, dockerContainer, task)
+	err = metadata.writeToMetadataFile(task, container, manager.cfg.DataDir)
+	if err != nil {
+		return err
+	}
+
+	// Update the file again with time stamp of last file update. This will cause the file to have
+	// a different time stamp from its actual file stat inspection but this is unavoidable in our
+	// design as we must populate the metadata with the update time before updating
+	updateTime, err := getMetadataFileUpdateTime(task, container, manager.cfg.DataDir)
+	if err != nil {
+		seelog.Errorf("Metadata file update time stamp not available: %s", err)
+	} else {
+		updateTimeFmt = updateTime.Format(time.StampNano)
+	}
+	metadata = manager.acquireMetadata(createTimeFmt, updateTimeFmt, dockerContainer, task)
```

Comment:
We need to populate the file with the last update time stamp but this is only possible by doing an update, collect the time stamp, then do another update with the new time stamp. This also means the file's stats will not match the recorded update time. We maintain the time stamp to serve as a log for the file's status.